state/applicationoffers_test.go

```diff
@@ -293,6 +293,62 @@ func (s *applicationOffersSuite) TestListOffersFilterOfferNameRegexp(c *gc.C) {
 	c.Assert(offers[0], jc.DeepEquals, offer)
 }
 
+func (s *applicationOffersSuite) TestListOffersAllowedUsersOwner(c *gc.C) {
+	sd := state.NewApplicationOffers(s.State)
+	offer, owner := s.createOffer(c, "offer1", "description for offer1")
+	s.createOffer(c, "offer2", "description for offer2")
+	s.createOffer(c, "offer3", "description for offer3")
+	offers, err := sd.ListOffers(crossmodel.ApplicationOfferFilter{
+		AllowedUsers: []string{owner, "mary"},
+	})
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(len(offers), gc.Equals, 1)
+	c.Assert(offers[0], jc.DeepEquals, offer)
+}
+
+func (s *applicationOffersSuite) TestListOffersAllowedUsers(c *gc.C) {
+	sd := state.NewApplicationOffers(s.State)
+	offer, _ := s.createOffer(c, "offer1", "description for offer1")
+	offer2, _ := s.createOffer(c, "offer2", "description for offer2")
+	s.createOffer(c, "offer3", "description for offer3")
+	s.Factory.MakeUser(c, &factory.UserParams{Name: "mary"})
+
+	mary := names.NewUserTag("mary")
+	err := s.State.CreateOfferAccess(
+		names.NewApplicationOfferTag(offer.OfferName), mary, permission.ConsumeAccess)
+	c.Assert(err, jc.ErrorIsNil)
+	err = s.State.CreateOfferAccess(
+		names.NewApplicationOfferTag(offer2.OfferName), mary, permission.ReadAccess)
+	c.Assert(err, jc.ErrorIsNil)
+	offers, err := sd.ListOffers(crossmodel.ApplicationOfferFilter{
+		AllowedUsers: []string{"mary"},
+	})
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(len(offers), gc.Equals, 1)
```

Comment:
This doesn't seem correct to me. You've created 3 offers, allowed Mary to access 2 of them, but ListOffers(AllowedUsers:mary) is only giving you 1 offer back?

Reply:
Getting 1 back is correct - the filtering is done on who has consume permission. The test sets up read permission as well to demonstrate that is filtered out